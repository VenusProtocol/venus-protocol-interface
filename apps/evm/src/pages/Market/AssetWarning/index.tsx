import { useState } from 'react';

import { useTranslation } from 'libs/translations';
import type { Pool, Token } from 'types';

import { Modal, Notice, TextButton } from 'components';
import { MarketTable } from 'containers/MarketTable';
import { useBreakpointUp } from 'hooks/responsive';
import TEST_IDS from './testIds';
import type { WarningType } from './types';

export interface AssetWarningProps extends React.HTMLAttributes<HTMLDivElement> {
  type: WarningType;
  token: Token;
  pool?: Pool;
  className?: string;
}

export const AssetWarning: React.FC<AssetWarningProps> = ({
  pool,
  token,
  type,
  className,
  ...otherProps
}) => {
  const [showAssets, setShowAssets] = useState(false);
  const { t, Trans } = useTranslation();
  const isSmOrUp = useBreakpointUp('sm');

  const translationArgs = {
    poolName: pool?.name,
    tokenSymbol: token.symbol,
  };

  const handleShowAssets = () => setShowAssets(true);
  const handleHideAssets = () => setShowAssets(false);

  if (!pool) {
    return null;
  }

  return (
    <div className={className} {...otherProps}>
      <Notice
        className="mb-2"
        description={
          <Trans
            // Translation key: do not remove this comment
            // t('assetWarning.borrowDescription')
            // t('assetWarning.supplyDescription')
            i18nKey={
              type === 'borrow'
                ? 'assetWarning.borrowDescription'
                : 'assetWarning.supplyDescription'
            }
            values={translationArgs}
            components={{
              Button: <TextButton className="p-0 h-auto font-medium" onClick={handleShowAssets} />,
            }}
          />
        }
      />

      <Modal
        isOpen={showAssets}
        handleClose={handleHideAssets}
        noHorizontalPadding={isSmOrUp}
        title={t('assetWarning.modalTitle', {
          poolName: pool.name,
        })}
      >
        <MarketTable
          data-testid={TEST_IDS.marketTable}
          rowOnClick={handleHideAssets}
          controls={false}
          selectVariant="quaternary"
          breakpoint="sm"
          cardClassName="p-0 pb-4 border-b-lightGrey rounded-none [&:not(:last-of-type)]:border-b-[1px] [&>hr]:h-0 [&>hr]:my-2"
          className="my-0 p-0 sm:p-0"
          pools={[pool]}
          columns={['asset', type === 'borrow' ? 'labeledBorrowApy' : 'supplyApy', 'liquidity']}
          initialOrder={{
            orderBy: type === 'borrow' ? 'labeledBorrowApy' : 'supplyApy',
            orderDirection: type === 'supply' ? 'desc' : 'asc',
          }}
        />
      </Modal>
    </div>
  );
};

export default AssetWarning;
