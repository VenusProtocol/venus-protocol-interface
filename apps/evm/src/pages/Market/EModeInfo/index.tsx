import { useTranslation } from 'libs/translations';
import type { EModeGroup, Token } from 'types';
import { areTokensEqual } from 'utilities';
import { MarketCard } from '../MarketCard';
import { AssetSettings, type AssetSettingsProps } from './AssetSettings';

export interface EModeInfoProps {
  eModeGroups: EModeGroup[];
  token: Token;
}

export const EModeInfo: React.FC<EModeInfoProps> = ({ eModeGroups, token }) => {
  const { t } = useTranslation();
  const assetSettingsProps = eModeGroups.reduce<Omit<AssetSettingsProps, 'isLast'>[]>(
    (acc, eModeGroup) => {
      const assetSettings = eModeGroup.assetSettings.find(settings =>
        areTokensEqual(token, settings.vToken.underlyingToken),
      );

      if (!assetSettings) {
        return acc;
      }

      const props: Omit<AssetSettingsProps, 'isLast'> = {
        eModeGroupName: eModeGroup.name,
        token: assetSettings.vToken.underlyingToken,
        collateralFactor: assetSettings.collateralFactor,
        liquidationThresholdPercentage: assetSettings.liquidationThresholdPercentage,
        liquidationPenaltyPercentage: assetSettings.liquidationPenaltyPercentage,
      };

      return [...acc, props];
    },
    [],
  );

  if (assetSettingsProps.length === 0) {
    return null;
  }

  return (
    <MarketCard title={t('asset.eModeInfo.title')}>
      <div className="space-y-4 sm:space-y-5">
        {assetSettingsProps.map((props, index) => (
          <AssetSettings
            {...props}
            isLast={index === assetSettingsProps.length - 1}
            key={props.eModeGroupName}
          />
        ))}
      </div>
    </MarketCard>
  );
};
