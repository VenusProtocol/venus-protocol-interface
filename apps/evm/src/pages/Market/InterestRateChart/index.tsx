import { cn, theme } from '@venusprotocol/ui';
import { useGetVTokenApySimulations } from 'clients/api';
import { ChartTooltipContent, Spinner } from 'components';
import { useTranslation } from 'libs/translations';
import {
  CartesianGrid,
  Line,
  LineChart,
  ReferenceLine,
  ResponsiveContainer,
  Tooltip,
  XAxis,
  YAxis,
} from 'recharts';
import type { Asset } from 'types';
import { formatPercentageToReadableValue } from 'utilities';
import { MarketCard, type MarketCardProps } from '../MarketCard';
import TEST_IDS from '../testIds';
import { useStyles } from './styles';

export interface InterestRateItem {
  utilizationRatePercentage: number;
  borrowApyPercentage: number;
  supplyApyPercentage: number;
}

export interface InterestRateChartProps {
  className?: string;
  asset: Asset;
  isIsolatedPoolMarket: boolean;
}

export const InterestRateChart: React.FC<InterestRateChartProps> = ({
  asset,
  className,
  isIsolatedPoolMarket,
}) => {
  const { t } = useTranslation();
  const styles = useStyles();

  const {
    isLoading: isInterestRateChartDataLoading,
    data: interestRateChartData = {
      apySimulations: [],
      currentUtilizationRatePercentage: 0,
    },
  } = useGetVTokenApySimulations({
    isIsolatedPoolMarket,
    asset,
  });

  const legends: MarketCardProps['legends'] = [
    {
      label: t('market.legends.utilizationRate'),
      color: 'blue',
    },
    {
      label: t('market.legends.borrowApy'),
      color: 'red',
    },
    {
      label: t('market.legends.supplyApy'),
      color: 'green',
    },
  ];

  return (
    <MarketCard
      className={className}
      data-testid={TEST_IDS.interestRateModel}
      title={t('market.interestRateModel.title')}
      legends={legends}
    >
      {isInterestRateChartDataLoading && interestRateChartData.apySimulations.length === 0 && (
        <Spinner />
      )}

      {interestRateChartData.apySimulations.length > 0 && (
        <div className="-mr-[10px]">
          <div className={cn('w-full h-62', className)}>
            <ResponsiveContainer>
              <LineChart
                data={interestRateChartData.apySimulations}
                margin={{
                  top: 20,
                  right: 10,
                  left: -12,
                }}
              >
                <XAxis
                  dataKey="utilizationRatePercentage"
                  axisLine={false}
                  tickLine={false}
                  tickFormatter={formatPercentageToReadableValue}
                  stroke={theme.colors.grey}
                  tickMargin={10}
                  tickCount={5}
                  type="number"
                  className="text-xs"
                />

                <YAxis
                  axisLine={false}
                  tickLine={false}
                  tickFormatter={formatPercentageToReadableValue}
                  tickMargin={10}
                  stroke={theme.colors.grey}
                  className="text-xs"
                  tickCount={10}
                />

                <Tooltip
                  isAnimationActive={false}
                  cursor={{ strokeDasharray: '4px 4px', stroke: theme.colors.grey }}
                  content={({ payload }) =>
                    payload?.[0] ? (
                      <ChartTooltipContent
                        items={[
                          {
                            label: t('interestRateChart.tooltipItemLabels.utilizationRate'),
                            value: formatPercentageToReadableValue(
                              payload?.[0].payload.utilizationRatePercentage,
                            ),
                          },
                          {
                            label: t('interestRateChart.tooltipItemLabels.borrowApy'),
                            value: formatPercentageToReadableValue(
                              payload?.[0].payload.borrowApyPercentage,
                            ),
                          },
                          {
                            label: t('interestRateChart.tooltipItemLabels.supplyApy'),
                            value: formatPercentageToReadableValue(
                              payload?.[0].payload.supplyApyPercentage,
                            ),
                          },
                        ]}
                      />
                    ) : null
                  }
                />

                <CartesianGrid vertical={false} stroke={theme.colors.lightGrey} />

                <Line
                  type="monotone"
                  dataKey="borrowApyPercentage"
                  stroke={styles.lineBorrowApyColor}
                  strokeWidth={2}
                  isAnimationActive={false}
                  activeDot={styles.lineActiveDot}
                  dot={false}
                />

                <Line
                  type="monotone"
                  dataKey="supplyApyPercentage"
                  stroke={styles.lineSupplyApyColor}
                  strokeWidth={2}
                  activeDot={styles.lineActiveDot}
                  isAnimationActive={false}
                  dot={false}
                />

                {typeof interestRateChartData.currentUtilizationRatePercentage === 'number' && (
                  <ReferenceLine
                    x={interestRateChartData.currentUtilizationRatePercentage}
                    stroke={styles.referenceLineColor}
                    // Note: we can not use the spread operator to extend
                    // styles.referenceLineLabel because its type is not accepted for
                    // that
                    label={Object.assign(styles.referenceLineLabel || {}, {
                      value: t('interestRateChart.currentUtilizationRateLabelValue', {
                        percentage: formatPercentageToReadableValue(
                          interestRateChartData.currentUtilizationRatePercentage,
                        ),
                      }),
                    })}
                    alwaysShow
                  />
                )}
              </LineChart>
            </ResponsiveContainer>
          </div>
        </div>
      )}
    </MarketCard>
  );
};
