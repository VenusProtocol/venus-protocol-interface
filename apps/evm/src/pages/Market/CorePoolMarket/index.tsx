/** @jsxImportSource @emotion/react */
import { useParams } from 'react-router-dom';

import { useGetChainMetadata } from 'hooks/useGetChainMetadata';

import MarketLoader from 'containers/MarketLoader';
import { Market } from '../Market';

const CorePoolMarket: React.FC = () => {
  const { vTokenAddress } = useParams();
  const { corePoolComptrollerContractAddress } = useGetChainMetadata();

  return (
    <MarketLoader
      poolComptrollerAddress={corePoolComptrollerContractAddress}
      vTokenAddress={vTokenAddress}
    >
      {marketProps => <Market {...marketProps} />}
    </MarketLoader>
  );
};

export default CorePoolMarket;
