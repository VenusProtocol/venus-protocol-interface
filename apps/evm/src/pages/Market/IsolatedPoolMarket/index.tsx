/** @jsxImportSource @emotion/react */
import { useParams } from 'react-router-dom';

import MarketLoader from 'containers/MarketLoader';
import { useIsFeatureEnabled } from 'hooks/useIsFeatureEnabled';
import { Market } from '../Market';
import MarketOld from '../MarketOld';

const IsolatedPoolMarket: React.FC = () => {
  const { vTokenAddress, poolComptrollerAddress } = useParams();
  const isNewMarketPageEnabled = useIsFeatureEnabled({ name: 'newMarketPage' });

  return (
    <MarketLoader
      poolComptrollerAddress={poolComptrollerAddress}
      vTokenAddress={vTokenAddress}
      isIsolatedPoolMarket
    >
      {marketProps =>
        isNewMarketPageEnabled ? <Market {...marketProps} /> : <MarketOld {...marketProps} />
      }
    </MarketLoader>
  );
};

export default IsolatedPoolMarket;
