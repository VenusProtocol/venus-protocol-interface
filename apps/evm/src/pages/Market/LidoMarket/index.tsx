import MarketLoader from 'containers/MarketLoader';
import { useGetChainMetadata } from 'hooks/useGetChainMetadata';
import { Market } from '../Market';

const LidoMarket: React.FC = () => {
  const { stakedEthPoolComptrollerContractAddress, wstEthContractAddress } = useGetChainMetadata();

  return (
    <MarketLoader
      poolComptrollerAddress={stakedEthPoolComptrollerContractAddress}
      vTokenAddress={wstEthContractAddress}
    >
      {marketProps => <Market {...marketProps} />}
    </MarketLoader>
  );
};

export default LidoMarket;
