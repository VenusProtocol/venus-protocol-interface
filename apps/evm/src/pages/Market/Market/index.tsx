import { AssetWarning, Card } from 'components';
import type { Asset, Pool } from 'types';

import { InterestRateChart } from './InterestRateChart';
import { MarketHistory } from './MarketHistory';
import MarketInfo from './MarketInfo';
import { OperationForm } from './OperationForm';

export interface MarketProps {
  asset: Asset;
  pool: Pool;
}

export const Market: React.FC<MarketProps> = ({ asset, pool }) => (
  <div className="py-6 md:py-8 xl:py-12">
    <AssetWarning
      token={asset.vToken.underlyingToken}
      pool={pool}
      type="supply"
      className="mb-4 sm:mb-6"
    />

    <div className="space-y-6 lg:flex lg:space-y-0 lg:gap-x-6">
      <Card className="w-auto self-start shrink-0 overflow-x-auto lg:order-2 lg:sticky lg:w-[400px] lg:top-6 lg:max-h-[calc(100vh-48px)]">
        <OperationForm poolComptrollerAddress={pool.comptrollerAddress} vToken={asset.vToken} />
      </Card>

      {/* w-0 is a hotfix to force the charts to adapt their size when resizing the window (see https://github.com/recharts/recharts/issues/172#issuecomment-307858843) */}
      <div className="space-y-6 lg:grow lg:order-1 lg:w-0">
        <MarketHistory asset={asset} poolComptrollerContractAddress={pool.comptrollerAddress} />

        <InterestRateChart asset={asset} isIsolatedPoolMarket={pool.isIsolated} />

        <MarketInfo asset={asset} />
      </div>
    </div>
  </div>
);
