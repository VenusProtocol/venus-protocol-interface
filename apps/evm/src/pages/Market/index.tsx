import { Paper } from '@mui/material';
import BigNumber from 'bignumber.js';
import { useMemo } from 'react';

import { useGetVTokenApySimulations } from 'clients/api';
import { Button, Card, SecondaryButton, Spinner } from 'components';
import { InterestRateChart, InterestRateChartProps } from 'components/charts/InterestRateChart';
import PLACEHOLDER_KEY from 'constants/placeholderKey';
import { useGetChainMetadata } from 'hooks/useGetChainMetadata';
import { useIsFeatureEnabled } from 'hooks/useIsFeatureEnabled';
import useIsTokenActionEnabled from 'hooks/useIsTokenActionEnabled';
import useOperationModal from 'hooks/useOperationModal';
import { useTranslation } from 'packages/translations';
import { Asset, Token } from 'types';
import {
  formatCentsToReadableValue,
  formatPercentageToReadableValue,
  formatTokensToReadableValue,
} from 'utilities';

import { MarketCard, MarketCardProps } from './MarketCard';
import { MarketHistory } from './MarketHistory';
import MarketInfo, { MarketInfoProps } from './MarketInfo';
import TEST_IDS from './testIds';

export interface MarketUiProps {
  interestRateChartData: InterestRateChartProps['data'];
  isInterestRateChartDataLoading: boolean;
  poolComptrollerAddress: string;
  currentUtilizationRatePercentage: number;
  asset: Asset;
  isBorrowActionEnabled: boolean;
  isSupplyActionEnabled: boolean;
  blocksPerDay: number;
}

export const MarketUi: React.FC<MarketUiProps> = ({
  asset,
  poolComptrollerAddress,
  isInterestRateChartDataLoading,
  interestRateChartData,
  currentUtilizationRatePercentage,
  isBorrowActionEnabled,
  isSupplyActionEnabled,
  blocksPerDay,
}) => {
  const { t } = useTranslation();

  const isMarketParticipantCountFeatureEnabled = useIsFeatureEnabled({
    name: 'marketParticipantCounts',
  });
  const isMarketHistoryFeatureEnabled = useIsFeatureEnabled({
    name: 'marketHistory',
  });

  const { openOperationModal, OperationModal } = useOperationModal();

  const { dailySupplyInterestsCents, dailyBorrowInterestsCents } = useMemo(
    () => ({
      // Calculate daily interests for suppliers and borrowers. Note that we don't
      // use BigNumber to calculate these values, as this would slow down
      // calculation a lot while the end result doesn't need to be extremely
      // precise

      // prettier-ignore
      dailySupplyInterestsCents: asset && +asset.supplyBalanceCents * (((1 + asset.supplyPercentageRatePerBlock.toNumber()) ** blocksPerDay) - 1),
      // prettier-ignore
      dailyBorrowInterestsCents: asset && +asset.borrowBalanceCents * (((1 + asset.borrowPercentageRatePerBlock.toNumber()) ** blocksPerDay) - 1),
    }),
    [asset, blocksPerDay],
  );

  const isSupplyOrBorrowEnabled = isSupplyActionEnabled || isBorrowActionEnabled;

  const interestRateModelLegends: MarketCardProps['legends'] = [
    {
      label: t('market.legends.utilizationRate'),
      color: 'blue',
    },
    {
      label: t('market.legends.borrowApy'),
      color: 'red',
    },
    {
      label: t('market.legends.supplyApy'),
      color: 'green',
    },
  ];

  const marketInfoStats: MarketInfoProps['stats'] = useMemo(() => {
    if (!asset) {
      return [];
    }

    const distributionMapping = [
      ...asset.supplyDistributions,
      ...asset.borrowDistributions,
    ].reduce<{
      [tokenSymbol: string]: {
        rewardToken: Token;
        dailyDistributedTokens: BigNumber;
      };
    }>((accDistributionMapping, distribution) => {
      // Filter out distributions that do not indicate the amount of tokens distributed to everyone
      // per day
      if (!('dailyDistributedTokens' in distribution)) {
        return accDistributionMapping;
      }

      const accCopy = { ...accDistributionMapping };

      if (!Object.prototype.hasOwnProperty.call(accCopy, distribution.token.address)) {
        accCopy[distribution.token.address] = {
          rewardToken: distribution.token,
          dailyDistributedTokens: new BigNumber(0),
        };
      }

      accCopy[distribution.token.address] = {
        ...accCopy[distribution.token.address],
        dailyDistributedTokens: accCopy[distribution.token.address].dailyDistributedTokens.plus(
          distribution.dailyDistributedTokens,
        ),
      };

      return accCopy;
    }, {});

    const distributionRows = Object.values(distributionMapping).map(
      ({ rewardToken, dailyDistributedTokens }) => ({
        label: t('market.marketInfo.stats.dailyDistribution', {
          tokenSymbol: rewardToken.symbol,
        }),
        value: formatTokensToReadableValue({
          value: dailyDistributedTokens,
          addSymbol: false,
          token: rewardToken,
        }),
      }),
    );

    const participantCountRows = isMarketParticipantCountFeatureEnabled
      ? [
          {
            label: t('market.marketInfo.stats.supplierCountLabel'),
            value: asset.supplierCount ?? '-',
          },
          {
            label: t('market.marketInfo.stats.borrowerCountLabel'),
            value: asset.borrowerCount ?? '-',
          },
        ]
      : [];

    return [
      {
        label: t('market.marketInfo.stats.priceLabel'),
        value: asset.tokenPriceCents
          ? formatCentsToReadableValue({
              value: asset.tokenPriceCents,
              isTokenPrice: true,
            })
          : PLACEHOLDER_KEY,
      },
      {
        label: t('market.marketInfo.stats.marketLiquidityLabel'),
        value: formatCentsToReadableValue({
          value: asset.liquidityCents,
        }),
      },
      ...participantCountRows,
      {
        label: t('market.marketInfo.stats.supplyCapLabel'),
        value: !asset.supplyCapTokens
          ? t('market.marketInfo.stats.unlimitedSupplyCap')
          : formatTokensToReadableValue({
              value: asset.supplyCapTokens,
              token: asset.vToken.underlyingToken,
            }),
      },
      {
        label: t('market.marketInfo.stats.borrowCapLabel'),
        value: !asset.borrowCapTokens
          ? t('market.marketInfo.stats.unlimitedBorrowCap')
          : formatTokensToReadableValue({
              value: asset.borrowCapTokens,
              token: asset.vToken.underlyingToken,
            }),
      },
      {
        label: t('market.marketInfo.stats.dailySupplyingInterestsLabel'),
        value: formatCentsToReadableValue({
          value: dailySupplyInterestsCents,
        }),
      },
      {
        label: t('market.marketInfo.stats.dailyBorrowingInterestsLabel'),
        value: formatCentsToReadableValue({
          value: dailyBorrowInterestsCents,
        }),
      },
      ...distributionRows,
      {
        label: t('market.marketInfo.stats.reserveTokensLabel'),
        value: formatTokensToReadableValue({
          value: asset.reserveTokens,
          token: asset.vToken.underlyingToken,
        }),
      },
      {
        label: t('market.marketInfo.stats.reserveFactorLabel'),
        value: formatPercentageToReadableValue(asset.reserveFactor && asset.reserveFactor * 100),
      },
      {
        label: t('market.marketInfo.stats.collateralFactorLabel'),
        value: formatPercentageToReadableValue(
          asset.collateralFactor && asset.collateralFactor * 100,
        ),
      },
      {
        label: t('market.marketInfo.stats.mintedTokensLabel', {
          vTokenSymbol: asset.vToken.symbol,
        }),
        value: formatTokensToReadableValue({
          value: asset.supplyBalanceTokens.multipliedBy(asset.exchangeRateVTokens),
          addSymbol: false,
          token: asset.vToken,
        }),
      },
      {
        label: t('market.marketInfo.stats.exchangeRateLabel'),
        value: asset.exchangeRateVTokens
          ? t('market.marketInfo.stats.exchangeRateValue', {
              tokenSymbol: asset.vToken.underlyingToken.symbol,
              vTokenSymbol: asset.vToken.symbol,
              rate: asset.exchangeRateVTokens.dp(6).toFixed(),
            })
          : PLACEHOLDER_KEY,
      },
    ];
  }, [
    asset,
    t,
    dailySupplyInterestsCents,
    dailyBorrowInterestsCents,
    isMarketParticipantCountFeatureEnabled,
  ]);

  const buttonsDom = (
    <div className="flex items-center space-x-4">
      {isSupplyActionEnabled && (
        <Button
          className="w-full"
          onClick={() =>
            openOperationModal({
              vToken: asset.vToken,
              poolComptrollerAddress,
              initialActiveTabIndex: 0,
            })
          }
        >
          {t('market.supplyButtonLabel')}
        </Button>
      )}
      {isBorrowActionEnabled && (
        <SecondaryButton
          className="w-full"
          onClick={() =>
            openOperationModal({
              vToken: asset.vToken,
              poolComptrollerAddress,
              initialActiveTabIndex: 2,
            })
          }
        >
          {t('market.borrowButtonLabel')}
        </SecondaryButton>
      )}
    </div>
  );

  return (
    <>
      <div className="space-y-6 xl:grid xl:grid-cols-3 xl:gap-8 xl:space-y-0">
        {isSupplyOrBorrowEnabled && <Card className="xl:hidden">{buttonsDom}</Card>}

        <div className="space-y-6 xl:col-span-2 xl:mt-0">
          {isMarketHistoryFeatureEnabled && <MarketHistory asset={asset} />}

          <MarketCard
            testId={TEST_IDS.interestRateModel}
            title={t('market.interestRateModel.title')}
            legends={interestRateModelLegends}
          >
            {isInterestRateChartDataLoading && interestRateChartData.length === 0 && <Spinner />}
            {interestRateChartData.length > 0 && (
              <div className="-mr-[10px]">
                <InterestRateChart
                  data={interestRateChartData}
                  currentUtilizationRatePercentage={currentUtilizationRatePercentage}
                />
              </div>
            )}
          </MarketCard>
        </div>

        <div className="xl:col-span-1 xl:space-y-6">
          {isSupplyOrBorrowEnabled && <Paper className="hidden xl:block">{buttonsDom}</Paper>}

          <MarketInfo stats={marketInfoStats} testId={TEST_IDS.marketInfo} />
        </div>
      </div>

      <OperationModal />
    </>
  );
};

interface MarketProps {
  poolComptrollerAddress: string;
  asset: Asset;
  isIsolatedPoolMarket: boolean;
}

const Market: React.FC<MarketProps> = ({
  poolComptrollerAddress,
  asset,
  isIsolatedPoolMarket = false,
}) => {
  const { blocksPerDay } = useGetChainMetadata();

  const {
    isLoading: isInterestRateChartDataLoading,
    data: interestRateChartData = {
      apySimulations: [],
      currentUtilizationRatePercentage: 0,
    },
  } = useGetVTokenApySimulations({
    vToken: asset.vToken,
    isIsolatedPoolMarket,
    asset,
  });

  const isBorrowActionEnabled = useIsTokenActionEnabled({
    tokenAddress: asset.vToken.underlyingToken.address,
    action: 'borrow',
  });

  const isSupplyActionEnabled = useIsTokenActionEnabled({
    tokenAddress: asset.vToken.underlyingToken.address,
    action: 'supply',
  });

  return (
    <MarketUi
      asset={asset}
      blocksPerDay={blocksPerDay}
      poolComptrollerAddress={poolComptrollerAddress}
      isInterestRateChartDataLoading={isInterestRateChartDataLoading}
      interestRateChartData={interestRateChartData.apySimulations}
      currentUtilizationRatePercentage={interestRateChartData.currentUtilizationRatePercentage}
      isBorrowActionEnabled={isBorrowActionEnabled}
      isSupplyActionEnabled={isSupplyActionEnabled}
    />
  );
};

export default Market;
