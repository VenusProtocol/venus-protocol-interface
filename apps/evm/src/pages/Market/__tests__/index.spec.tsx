import { waitFor } from '@testing-library/react';
import type { Mock } from 'vitest';

import { poolData } from '__mocks__/models/pools';
import { useGetAsset, useGetPool } from 'clients/api';
import { renderComponent } from 'testUtils/render';
import Market from '..';

vi.mock('react-router', async importOriginal => {
  const actual: Record<string, unknown> = await importOriginal();

  return {
    ...actual,
    useParams: () => ({
      vTokenAddress: '0xfakeVTokenAddress',
      poolComptrollerAddress: '0xfakePoolComptrollerAddress',
    }),
  };
});

describe('Market', () => {
  beforeEach(() => {
    (useGetPool as Mock).mockImplementation(() => ({
      isLoading: false,
      data: {
        pool: poolData[0],
      },
    }));

    (useGetAsset as Mock).mockImplementation(() => ({
      isLoading: false,
      data: {
        asset: poolData[0].assets[0],
      },
    }));
  });

  it('displays content correctly', async () => {
    const { container } = renderComponent(<Market />);

    await waitFor(() => expect(container.textContent).not.toEqual(''));

    expect(container.textContent).toMatchSnapshot();
  });
});
