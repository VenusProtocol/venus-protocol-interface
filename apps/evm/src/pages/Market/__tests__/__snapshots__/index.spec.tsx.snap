// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Market > displays content correctly 1`] = `"Supplying XVS to the Venus pool will enable you to borrow tokens from this pool exclusively. Show all marketsSupplyWithdrawBorrowRepayCollateralMAXSupply APY0.05%Distribution APY (XVS)0.12%Total APY0.17%Connect walletSupply info0%Total supplied> $100T / > $100T> 100T / > 100T XVSCurrent APY0.17%Borrow info< 0.01%Total borrowed$2.36M / > $100T1.85M / > 100T XVSCurrent APY-6.48%Liquidation Threshold50%Liquidation Penalty10%Interest Rate ModelUtilization rateBorrow APYSupply APYE-mode infoStablecoinsMax LTV60%Liquidation threshold62%Liquidation penalty0%Market infoDaily supplying interests$3.98Daily borrowing interests-$44.81Daily XVS distributed19.99MReserve factor25%Exchange rate1 XVS=49.589181 vXVS"`;
